src/Grpc.Net.Client/Internal/Retry/HedgingCall.cs

```diff
@@ -0,0 +1,404 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Grpc.Core;
+using Grpc.Net.Client.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Grpc.Net.Client.Internal.Retry
+{
+    internal sealed partial class HedgingCall<TRequest, TResponse> : RetryCallBase<TRequest, TResponse>
+        where TRequest : class
+        where TResponse : class
+    {
+        // Getting logger name from generic type is slow. Cached copy.
+        private const string LoggerName = "Grpc.Net.Client.Internal.HedgingCall";
+
+        private readonly HedgingPolicyInfo _hedgingPolicy;
+
+        private int _callsAttempted;
+
+        private CancellationTokenSource? _hedgingDelayCts;
+        private TaskCompletionSource<object?>? _delayInterruptTcs;
+        private TimeSpan? _pushbackDelay;
+
+        // Internal for testing
+        internal List<IGrpcCall<TRequest, TResponse>> _activeCalls { get; }
+        internal Task? CreateHedgingCallsTask { get; set; }
+
+        public HedgingCall(HedgingPolicyInfo hedgingPolicy, GrpcChannel channel, Method<TRequest, TResponse> method, CallOptions options)
+            : base(channel, method, options, LoggerName, hedgingPolicy.MaxAttempts)
+        {
+            _hedgingPolicy = hedgingPolicy;
+            _activeCalls = new List<IGrpcCall<TRequest, TResponse>>();
+
+            if (_hedgingPolicy.HedgingDelay > TimeSpan.Zero)
+            {
+                _delayInterruptTcs = new TaskCompletionSource<object?>(TaskCreationOptions.None);
```

Comment:
Is this intentionally None and not RunContinuationsAsynchronously?
```suggestion
                _delayInterruptTcs = new TaskCompletionSource<object?>(TaskCreationOptions.RunContinuationsAsynchronously);
```

You're completing the Tcs in a lock and can then run the callback inline.

Reply:
This was intentional at one point but it can be continued async now.